modules/rostests/kmtests/ntos_ob/ObQuery.c

```diff
@@ -0,0 +1,113 @@
+/*
+ * PROJECT:     ReactOS kernel-mode tests
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Kernel mode tests for object information querying
+ * COPYRIGHT:   Copyright 2023 George Bișoc <george.bisoc@reactos.org>
+ */
+
+#include <kmt_test.h>
+
+#define OBJ_REGISTRY_TYPE_INFO_SIZE   104
+#define OBJ_GLOBAL_DIR_TYPE_INFO_SIZE 116
+
+static
+VOID
+ObjectBasicInformationTests(VOID)
+{
+    NTSTATUS Status;
+    HANDLE KeyHandle;
+    HANDLE GlobalDirHandle;
+    OBJECT_BASIC_INFORMATION BasicInfo;
+    ULONG ReturnLength;
+    OBJECT_ATTRIBUTES ObjectAttributes;
+    static UNICODE_STRING GlobalDir = RTL_CONSTANT_STRING(L"\\??");
+    static UNICODE_STRING RegRootKey = RTL_CONSTANT_STRING(L"\\Registry");
+
+    /* We must be in PASSIVE_LEVEL to do all of this stuff */
+    ok_irql(PASSIVE_LEVEL);
+
+    /* Create a key path to the registry root */
+    InitializeObjectAttributes(&ObjectAttributes,
+                               &RegRootKey,
+                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
+                               NULL,
+                               NULL);
+    Status = ZwOpenKey(&KeyHandle, KEY_READ, &ObjectAttributes);
+    if (!NT_SUCCESS(Status))
+    {
+        ok(FALSE, "Failed to open \\Registry (Status 0x%lx)\n", Status);
+        return;
+    }
+
+    /* Give 0 as information length, this must fail */
+    Status = ZwQueryObject(KeyHandle,
+                           ObjectBasicInformation,
+                           &BasicInfo,
+                           0,
+                           &ReturnLength);
+    ok_eq_hex(Status, STATUS_INFO_LENGTH_MISMATCH);
+
+    /* Do a proper query now */
+    Status = ZwQueryObject(KeyHandle,
+                           ObjectBasicInformation,
+                           &BasicInfo,
+                           sizeof(BasicInfo),
+                           &ReturnLength);
+    ok_eq_hex(Status, STATUS_SUCCESS);
+
+    /* These must not be 0 because the registry is currently used */
+    ok(BasicInfo.HandleCount != 0, "\\Registry is in use but HandleCount is 0!\n");
+    ok(BasicInfo.PointerCount != 0, "\\Registry is in use but PointerCount is 0!\n");
+
+    /*
+     * NameInfoSize must be 0 because \Registry does not have
```

Comment:
Isn't an object directory an object itself with a name? If so, why isn't this counted in the nameinfosize?

Reply:
> Isn't an object directory an object itself with a name? If so, why isn't this counted in the nameinfosize?

Technically yes, an object that has a name provided by the object attributes should be included in the name info but registry keys are special in this regard, as `CmpParseKey` (the registry parser) doesn't take into account the object attributes during an open or create procedure. So whenever a key is parsed like here, we allocate a key body object that is nameless. The Windows' behavior is similar as ours, which is why that test doesn't fail. I guess I should probably update the comment to give an explanation about this.

See those `ObCreateObject` calls in the parser:
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/config/cmparse.c;h=52a3219babb72077cd3cca42dc610408e1254da3;hb=HEAD#l270
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/config/cmparse.c;h=52a3219babb72077cd3cca42dc610408e1254da3;hb=HEAD#l682

So basically, as a TL;DR...

- Initialize object attributes, with `\Registry` as object name path
- Open the key with `ZwOpenKey`
- Go all the way down until `CmpParseKey`
- Parse the key, open it with `CmpDoOpen`
- Allocate a key body object with `ObCreateObject`, no object attributes given
- Query basic information with `ZwQueryObject`
- `NameInfoSize == 0`

Now the question is, does `\Registry` have a name if at all? It does, when the registry root is create we don't invoke the registry parser to do it but instead we directly call OB to do the job for us.
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/config/cmsysini.c;hb=7c14c0ed7346c81311d8179562d489439273dca1#l1090

Reply:
The `\Registry` object has a name, and it's `\Registry` -- you can query `ObjectNameInformation` (or call `ObQueryNameString`) to find out. I believe `NameInfoSize` is intended to give the length of the buffer required for `ObjectNameInformation`, except that for some reason it just returns 0 for objects in the root directory?

Reply:
Apparently querying the name with `ObjectNameInformation` does indeed return you the name of the registry object, I guess the discussion in regard of the registry parser above is irrelevant at this point. So I dunno, I'm out of ideas, all I can say is that I am getting a NULL `NameInfo` when grabbing it from the object header, hence `NameInfoSize` will become 0. Everyone is free to volunteer why does it behave like that.